### ISSUE TITLE
Issues on VBx reimplementation

### ISSUE THREAD

**Issue by alephpi** [BUTSpeechFIT/VBx#67]
Mel-filterbank implementation differs from `librosa.filters.mel`
Vbx
```python
fbank_mx = features.mel_fbank_mx(winlen=400, fs=16000, NUMCHANS=64, LOFREQ=20.0, HIFREQ=7600, htk_bug=False)
```

librosa (version 0.10.1) (`n_fft=512` adapt to `winlen=400` for param `nfft` in `mel_fbank_mx` is computed as the next higher power of 2)
```python
fbank_mx_alt = librosa.filters.mel(sr=16000, n_fft=512,n_mels=64, fmin=20.0, fmax=7600, norm=None)
```
Here is two visualizations of the filterbanks with the following scripts (results from librosa implementation is transposed to adapt the shape of yours, i.e.(257, 64))
```python
import numpy as np
import matplotlib.pyplot as plt

fig, ax = plt.subplots()
img = librosa.display.specshow(fbank_mx, x_axis='linear', ax=ax)
ax.set(ylabel='Mel filter', title='Mel filter bank-Vbx')
fig.colorbar(img, ax=ax)

fig, ax = plt.subplots()
img = librosa.display.specshow(fbank_mx_alt.T, x_axis='linear', ax=ax)
ax.set(ylabel='Mel filter', title='Mel filter bank-librosa')
fig.colorbar(img, ax=ax)
```
![98cae8ca-15a9-4a40-b891-5c3a8547f883](https://github.com/BUTSpeechFIT/VBx/assets/61275421/84c2aca5-9d5a-4fc0-9617-922afa74b242)

![14e28081-7de4-448c-adea-2f3e173ba64b](https://github.com/BUTSpeechFIT/VBx/assets/61275421/3bc08dbb-1b3c-46fc-aca8-1a13032ff3ff)

I find the major difference is librosa's implementation has a linear scale in beginning while yours is always in log-scale. In code:
https://github.com/BUTSpeechFIT/VBx/blob/57466e6e245d5cdfe2e88ee6503702ace3ffdd03/VBx/features.py#L20-L21

https://github.com/librosa/librosa/blob/2f20cb53d0d73f631aa4f42e2448178e13f0f7e3/librosa/core/convert.py#L1255-L1307

Could you explain such difference? Thank you!

**Comment by alephpi** [comment]
Ah I find that I should set `htk=True` i.e.
```python
fbank_mx_alt = librosa.filters.mel(sr=16000, n_fft=512, n_mels=64, fmin=20.0, fmax=7600, norm=None, htk=True)
```
This way, the relative error is less than 2%, which I guess is ok for replacing your implementation by librosa's, do you agree?

**Comment by alephpi** [comment]
Hi, I came to realize this `feature.py` may be adapted from kaldi, right? Since I find [`torchaudio.compliance.kaldi.fbank()`](https://pytorch.org/audio/stable/_modules/torchaudio/compliance/kaldi.html#fbank) and its functionality (as well as the parameters) is very similar to your scripts. (I want to refactor VBx codes with as few as possible customized scripts and make it cleaner)
the following is what I configured by guessing, but I cannot replicate the same results
 ```python
fea_alt = fbank(segment, dither=8, frame_length=frame_length, frame_shift=frame_shift, 
      low_freq=20, high_freq=7600, min_duration=0.01, num_mel_bins=64, htk_compat=True, subtract_mean=True)
```
**If this is similar to kaldi, how should I set these parameters?**

here is the full scripts comparing the two, maybe helpful
```python
from pyannote.database.util import load_lab, load_rttm, load_lst
import soundfile
import numpy as np
import torchaudio
uri = 'ES2005a'
lab = load_lab(f'./example/vad/{uri}.lab', uri=uri)
annotation = load_rttm(f'./example/rttm/{uri}.rttm')[uri]
waveform, sr = torchaudio.load(f'./example/audios/16k/{uri}.wav')
signal, sr = soundfile.read(f'./example/audios/16k/{uri}.wav')
import features
if sr == 8000:
    noverlap = 120
    winlen = 200
    window = features.povey_window(winlen)
    fbank_mx = features.mel_fbank_mx(
        winlen, sr, NUMCHANS=64, LOFREQ=20.0, HIFREQ=3700, htk_bug=False)
elif sr == 16000:
    noverlap = 240
    winlen = 400
    window = features.povey_window(winlen)
    fbank_mx = features.mel_fbank_mx(
        winlen, sr, NUMCHANS=64, LOFREQ=20.0, HIFREQ=7600, htk_bug=False)
else:
    raise ValueError(f'Only 8kHz and 16kHz are supported. Got {sr} instead.')

LC = 150
RC = 149

np.random.seed(3)  # for reproducibility
# dithering, add small noises to signal to reduce artifacts
signal +=  8 * (np.random.rand(*signal.shape)*2 - 1)

for seg in lab.itersegments():
    if seg.duration > 0.01: # process segment only if longer than 0.01s
        start, end = int(seg.start * sr), int(seg.end * sr)
        segment = signal[start:end]
        segment_alt = waveform[:, start:end]
        # Mirror noverlap//2 initial and final samples
        segment = np.r_[segment[noverlap // 2 - 1::-1],
                        segment, segment[-1:-winlen // 2 - 1:-1]]
        fea = features.fbank_htk(segment, window, noverlap, fbank_mx,
                                        USEPOWER=True, ZMEANSOURCE=True)
        fea = features.cmvn_floating_kaldi(fea, LC, RC, norm_vars=False).astype(np.float32)

        frame_length = winlen / sr * 1000
        frame_shift = (winlen-noverlap) / sr * 1000
        from torchaudio.compliance.kaldi import fbank
        fea_alt = fbank(segment_alt, dither=8, frame_length=frame_length, frame_shift=frame_shift, 
            low_freq=20, high_freq=7600, min_duration=0.01, num_mel_bins=64, htk_compat=True, subtract_mean=True)
        break
```

**Comment by fnlandini** [comment]
Hi,
This code was originally written in C by Lukas Burget and the initial version was made to be compatible with HTK (hence, the flag you mentioned). Then, Kaldi took after that version of this feature extraction and this python code was later updated to match Kaldi features.
I suggest that you remove the dithering parts to compare the features. Dithering adds a bit of noise and then the results will always differ. Still, even if they differ, the final outcome of the whole pipeline might still be similar.
In any case, we are not sure what is the goal with this exercise. We are not sure if the current implementation can be sped up much more with this modification since most of the computation happens inside the x-vector extractor.

**Comment by alephpi** [comment]
ok I will try. Such refactor simply aims to adapt to my own coding taste.

**Comment by alephpi** [comment]
When I comment out the dithering line in your original code, i.e. https://github.com/BUTSpeechFIT/VBx/blob/57466e6e245d5cdfe2e88ee6503702ace3ffdd03/VBx/predict.py#L163-L164
**I get `np.allclose(fea, 0)` for most of the segments, I print out the log**
```python
# VBx/predict.py
...
# np.random.seed(3)  # for reproducibility
# signal = features.add_dither((signal*2**15).astype(int))

for segnum in range(len(labs)):
    seg = signal[labs[segnum, 0]:labs[segnum, 1]]
    print(f'segments {labs[segnum, 0]/samplerate}s to {labs[segnum, 1]/samplerate}s')
    if seg.shape[0] > 0.01*samplerate:  # process segment only if longer than 0.01s
        # Mirror noverlap//2 initial and final samples
        seg = np.r_[seg[noverlap // 2 - 1::-1],
                    seg, seg[-1:-winlen // 2 - 1:-1]]
        fea = features.fbank_htk(seg, window, noverlap, fbank_mx,
                                 USEPOWER=True, ZMEANSOURCE=True)
        fea = features.cmvn_floating_kaldi(fea, LC, RC, norm_vars=False).astype(np.float32)
        print(f'fea is all close to 0: {np.allclose(fea, 0)}')
        slen = len(fea)
        start = -seg_jump
...
```
Here is the log, can you verify? Is it normal? BTW if I set dither to 0 in `torchaudio.compliance.kaldi.fbank()`, I still get non-trivial results.
```
segments 0.0s to 25.15s
fea is all close to 0: True
segments 25.52s to 34.49s
fea is all close to 0: True
segments 35.9s to 37.45s
fea is all close to 0: True
segments 38.33s to 40.4s
fea is all close to 0: True
segments 40.42s to 106.7s
fea is all close to 0: False
segments 106.71s to 133.66s
fea is all close to 0: False
segments 133.88s to 136.26s
fea is all close to 0: True
segments 137.19s to 142.76s
fea is all close to 0: True
segments 143.04s to 179.41s
fea is all close to 0: False
segments 180.43s to 198.39s
fea is all close to 0: True
segments 199.71s to 200.04s
fea is all close to 0: True
segments 201.71s to 214.21s
fea is all close to 0: True
segments 214.86s to 215.16s
fea is all close to 0: True
segments 215.21s to 224.01s
fea is all close to 0: True
segments 225.08s to 227.0s
fea is all close to 0: True
segments 228.96s to 240.98s
fea is all close to 0: True
segments 241.37s to 242.33s
fea is all close to 0: True
segments 243.63s to 248.52s
fea is all close to 0: False
segments 250.2s to 251.79s
fea is all close to 0: True
segments 253.1s to 254.65s
fea is all close to 0: True
segments 255.08s to 256.06s
fea is all close to 0: True
segments 266.84s to 269.27s
fea is all close to 0: True
segments 276.78s to 283.19s
fea is all close to 0: True
segments 283.87s to 301.8s
fea is all close to 0: False
segments 302.14s to 306.59s
fea is all close to 0: True
```

**Comment by alephpi** [comment]
When you do dithering as 
https://github.com/BUTSpeechFIT/VBx/blob/57466e6e245d5cdfe2e88ee6503702ace3ffdd03/VBx/predict.py#L163-L164
what is the purpose of `signal*2**15.astype(int)`? I find if I comment out this whole line, the `fea` will be trivial, as in the previous hidden comment (I hide it since I thought it was due to the noise but I realize it seems due to the scaling)
I guess it's something related to signal processing tricks, but sorry I don't know it since I can just drop the raw signal into `torchaudio.compliance.kaldi.fbank()` and it works...

**UPDATE**: refer to https://github.com/pytorch/audio/issues/371#issuecomment-625613872

**Comment by fnlandini** [comment]
It puts the values into a larger range so that the other operations that are applied later make sense with the given parameters. In particular, the dithering (with the current parameters) has to be applied after rescaling. Otherwise, the introduced noise will be too large compared to the pre-dithering values.

**Comment by alephpi** [comment]
The ratio 100 means one frame in mel filterbank corresponds to 10ms(which is the shift the window) right? i.e. the mel frame has a samplerate of 100 frames per seconds, right? 
```python
seg_end = round( 
             labs[segnum, 0] / float(samplerate) + start / 100.0 + seg_len / 100.0, 3 
         )
 ```
https://github.com/BUTSpeechFIT/VBx/blob/57466e6e245d5cdfe2e88ee6503702ace3ffdd03/VBx/predict.py#L179-L193

**Comment by fnlandini** [comment]
Yes, that's correct. The window length is 25ms and they overlap 15ms so the framerate is 1 frame every 10ms (100 per second)

**Comment by alephpi** [comment]
Hi, I just wonder in your AHC baseline, it seems you only use cosine similarity metric rather than plda scores to cluster the speaker embeddings, right? I notice the plda scoring function but find it applies nowhere. 
https://github.com/BUTSpeechFIT/VBx/blob/57466e6e245d5cdfe2e88ee6503702ace3ffdd03/VBx/diarization_lib.py#L59-L93

**Comment by fnlandini** [comment]
Hi, in the current version there is cosine similarity but in previous versions we had [kaldi_ivector_plda_scoring_dense](https://github.com/BUTSpeechFIT/VBx/blob/7103b76a32dec65239d63fd8047496a71dd0602e/VBx/vbhmm.py#L123)) .
It made sense to keep the function in case anyone would rather use it. Having the PLDA allows to do adaptation like we did [here](https://www.fit.vutbr.cz/research/groups/speech/publi/2020/diez_icassp2020_09053982.pdf)

**Comment by fnlandini** [comment]
Closing due to inactivity
